pkgs/applications/window-managers/herbstluftwm/default.nix

```diff
@@ -1,24 +1,26 @@
 { stdenv, fetchurl, cmake, pkgconfig, glib, libX11, libXext, libXinerama, libXrandr
-, withDoc ? stdenv.buildPlatform == stdenv.targetPlatform, asciidoc ? null }:
+, python3, xdotool, xorg, xterm, runtimeShell
+, withDoc ? stdenv.buildPlatform == stdenv.targetPlatform, asciidoc ? null
```

Comment:
```suggestion
, withDoc ? stdenv.buildPlatform == stdenv.targetPlatform, asciidoc
```

Reply:
From what I see in nixpkgs, optional dependencies (enabled with withX, enableX, Xsupport, etc) are defaulted to null to avoid throwing an error and an assertion is added. See [pkgs/applications/video/mplayer/default.nix](https://github.com/NixOS/nixpkgs/blob/4b610d814e09857c974ac2f65374a5baba0902ff/pkgs/applications/video/mplayer/default.nix) for instance. Unless this has changed recently?

Reply:
This is just syntactic sugar and is not required if you use lib.optionals.

Reply:
No, it is not syntactic sugar. The semantic differs. `asciidoc ? null` ensures the derivation can be evaluated *even* if asciidoc has not been found by `callPackage`. When asciidoc is not needed we must not enforce this false dependency. Thus the common pattern is :
- defaulting asciidoc to `null`, so `callPackage` may not find it : this is not an error.
- ensuring we actually have asciidoc *only* when needed, with an assertion.
(I agree that, in this specific case and *for now*, asciidoc is overridden in all-packages.nix anyway.)
This is a common pattern in nixpkgs and it makes sense.

Reply:
> No, it is not syntactic sugar.

It is. Not required. At all. See how frameworks are handled on darwin. Nothing of the frameworks evals on linux but trough the usage of lib.opional(s) we do not evaluate it.

> asciidoc ? null ensures the derivation can be evaluated even if asciidoc has not been found by callPackage

> defaulting asciidoc to null, so callPackage may not find it : this is not an error.

That is wrong. The variable only needs to be defined. It can even be defined to null and evaluating won't fail.

I even gave this a try but a dependency wouldn't build with it but it evaluated.

> When asciidoc is not needed we must not enforce this false dependency.

This is only if doc is build. Otherwise nothing happens. asciidoc is only build and used by your package if withDoc is true. If it is nill or false and I removed all your extra code it is still not build for asciidoc.

> ensuring we actually have asciidoc only when needed, with an assertion.

That is not the case. Right now you are checking asciidoc is not overwritten with null.

If it would be null than you couldn't build herbstluftwm at all because ninja would fail to build. This proves you did not properly test this flag.

> I agree that, in this specific case and for now, asciidoc is overridden in all-packages.nix anyway.

It is not overwritten but the inputs is assigned a different version. It does not influence any of the things in here and makes no difference.

> This is a common pattern in nixpkgs and it makes sense.

It is a relic of the past and in most places we could remove it without any impact.

To prove my point I build this PR with the below diff without any eval errors:

```diff
diff --git a/pkgs/applications/window-managers/herbstluftwm/default.nix b/pkgs/applications/window-managers/herbstluftwm/default.nix
index 17a9784a339..ab5280d9686 100644
--- a/pkgs/applications/window-managers/herbstluftwm/default.nix
+++ b/pkgs/applications/window-managers/herbstluftwm/default.nix
@@ -1,14 +1,10 @@
 { stdenv, fetchurl, cmake, pkgconfig, python3, glib, libX11, libXext, libXinerama, libXrandr
 , xdotool, xorgserver, xsetroot, xterm, runtimeShell
-, withDoc ? stdenv.buildPlatform == stdenv.targetPlatform, asciidoc ? null
+, withDoc ? false
 , nixosTests }:
 
-# Doc generation is disabled by default when cross compiling because asciidoc
-# dependency is broken when cross compiling for now
-
-assert withDoc -> asciidoc != null;
-
 stdenv.mkDerivation rec {
+  asciidoc = null;
   pname = "herbstluftwm";
   version = "0.9.1";
 
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index fa9f0fe1de5..1eb4686fae1 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -22164,7 +22164,6 @@ in
   herbe = callPackage ../applications/misc/herbe { };
 
   herbstluftwm = callPackage ../applications/window-managers/herbstluftwm {
-    asciidoc = asciidoc-full;
   };
 
   hercules = callPackage ../applications/virtualization/hercules { };
```


But it failed to build because the the doc tests ``../tests/test_doc.py`` failed.

So please accept my suggestions and fix the flag you wanted to introduce because right now it is broken.

Reply:
I think you missed my point: this is about the `callPackage` mechanism. *Not the evaluation nor the realisation*. Consider this:
```
> callPackage ({ doesNotExist, hello }: hello) {}
error: anonymous function at (string):1:14 called without required argument 'doesNotExist'
```
You'll get this error while `doesNotExist` *is* a false dependency. So things *happens*.

I will not answer all of your comments (but I can if you want) because I think we are both wasting time since we did not discussed the same thing. You try to demonstrate something which is not about what I talked about. (Some sentences show that you clearly did not understand what I mean – I apologize if I was not clear enough. Your diff removes the assertion that asciidoc is not null, set it to null and you say that the tests fail. Obviously. But what are you trying to prove?)

That being said, I'm willing to do the change (as I said "unless this has changed recently?") if this is the current pattern in nixpkgs (even though I think that's not a good thing and is a loss of information). Could you just direct me to a discussion/PR/issue/comment/RFC stating that? Thanks.

Reply:
> But what are you trying to prove?

That it is not required and your assertion does nothing and we can save those characters.

Also you option is broken and does not work despite you having extra code which does nothing.

> You'll get this error while doesNotExist is a false dependency. So things happens.

If asciidoc is not defined ninja will not build which is required in the dependency tree of this one.

> not a good thing and is a loss of information

Having an option in inputs and the lib.optional is enough information for everyone to know that those packages are required if you want to enable that option.

> Could you just direct me to a discussion/PR/issue/comment/RFC stating that?

The github search is terrible and I will never find anything relevant in there. Assertion in packages are mostly useful when you have multiple options which can't be activated at once.

Packages that do it like this are mostly old and arguing that a lot of packages do it will never yield any change in this regard.